src/backend/cdb/cdbmutate.c

```diff
@@ -788,6 +809,106 @@ apply_motion_mutator(Node *node, ApplyMotionState *context)
 			return plan_tree_mutator(node, apply_motion_mutator, context);
 	}
 
+	/*
+	 * For UPDATE/DELETE, we check if there's any motions before scan in the
+	 * same subtree for the table we're going to modify. If we encounter the
+	 * scan before any motions, then we can elide unneccessary Explicit
+	 * Redistribute Motion.
+	 */
+	if (IsA(node, ModifyTable))
+	{
+		ModifyTable *mt = (ModifyTable *) node;
+
+		if (mt->operation == CMD_UPDATE || mt->operation == CMD_DELETE)
+		{
+			ListCell   *lcr;
+
+			/*
+			 * Sanity check, since we don't allow multiple ModifyTable nodes.
+			 */
+			Assert(context->mt.resultRtis == NULL);
+			Assert(context->mt.nMotionsAbove == 0);
+			Assert(!context->mt.needExplicitMotion);
+			Assert(!context->mt.isChecking);
+
+			/*
+			 * When UPDATE/DELETE occurs on a partitioned table, or a table that
+			 * is a part of inheritance tree, ModifyTable node will have more
+			 * than one relation in resultRelations.
+			 *
+			 * We make a set of resulting relations' indexes to compare them
+			 * later.
+			 */
+			foreach(lcr, mt->resultRelations)
+			{
+				context->mt.resultRtis = bms_add_member(context->mt.resultRtis,
+														lfirst_int(lcr));
+			}
+
+			context->mt.isChecking = true;
+		}
+	}
+
+	/*
+	 * Elide Explicit Redistribute Motion if there's no motions between the
+	 * scan and the ModifyTable.
+	 */
+	if (context->mt.isChecking)
+	{
+
+		/* Remember if we are descending into a motion node. */
+		if (IsA(node, Motion))
+			context->mt.nMotionsAbove += 1;
+
+		/*
+		 * If this is a scan and it's scanrelid matches ModifyTable's relid,
+		 * we encountered a scan before any Motions.
```

Comment:
> we encountered a scan before any Motions.

it sounds like we get into the next block only if we haven't encountered motion, which is not true.